contract_compassion/model/contracts.py

```diff
@@ -0,0 +1,646 @@
+
+    def on_change_next_invoice_date(
+            self, cr, uid, ids, new_invoice_date, context=None):
+        res = True
+        for contract in self.browse(cr, uid, ids, context):
+            if (contract.state not in ('draft', 'mandate')):
+                res = super(
+                    self._name, self).on_change_next_invoice_date(
+                        self, cr, uid, ids, new_invoice_date, context) & res
+        return res
+
+    ################################
+    #        FIELDS METHODS        #
+    ################################
+
+    def _active(self, cr, uid, ids, field_name, args, context=None):
+        # Dummy function that sets the active flag.
+        self._on_contract_active(cr, uid, ids, context=context)
+        return {id: True for id in ids}
+
+    def _get_contract_from_invoice(invoice_obj, cr, uid, invoice_ids,
+                                   context=None):
+        self = invoice_obj.pool.get('recurring.contract')
+        res = set()
+        # Read data in english
+        if context is None:
+            context = {}
+        ctx = context.copy()
+        ctx['lang'] = 'en_US'
+        for invoice in invoice_obj.browse(cr, uid, invoice_ids, ctx):
+            if invoice.state == 'paid':
+                self._invoice_paid(cr, uid, invoice, ctx)
+
+                pay_dates = [move_line.date
+                             for move_line in invoice.payment_ids
+                             if move_line.credit > 0] or [0]
+
+                last_pay_date = max(pay_dates)
+                first_pay_date = min(pay_dates)
+
+                for invoice_line in invoice.invoice_line:
+                    contract = invoice_line.contract_id
+
+                    if contract.id not in res and (
+                            contract.state == 'waiting' and last_pay_date):
+                        # Activate the contract and set the
+                        # activation_date
+                        res.add(contract.id)
+                        contract.write({
+                            'activation_date': datetime.today().strftime(DF)})
+
+                        # Cancel the old invoices if a contract is activated
+                        self._cancel_old_invoices(
+                            cr, uid,
+                            invoice.partner_id.id,
+                            contract.id,
+                            first_pay_date,
+                            context)
+        return list(res)
+
+    def _cancel_old_invoices(
+            self, cr, uid, partner_id,
+            contract_id, date_invoice, context=None):
+        '''
+            Cancel the invoices of a partner from a date
+            If the invoice has only one contract -> cancel
+            Else -> draft to modify the invoice and validate
+        '''
+        invoice_line_obj = self.pool.get('account.invoice.line')
+        invoice_obj = self.pool.get('account.invoice')
+        invoice_ids = invoice_obj.search(

```

Comment:
Instead of starting with invoices, you could directly search for invoice_lines (contract_id = contract_id and state = open and due_date < date_invoice).